src/HttpUpgrade.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 1996-2019 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_HTTP_UPGRADE_H
+#define SQUID_HTTP_UPGRADE_H
+
+#include "acl/forward.h"
+#include "base/StringView.h"
+#include "sbuf/forward.h"
+
+#include <map>
+
+/// a reference to a protocol name[/version] string; no 0-termination is assumed
+class ProtocolView
+{
+public:
+    ProtocolView(const char * const start, const size_t len);
+    explicit ProtocolView(const StringView &proto);
+    explicit ProtocolView(const SBuf &proto);
+
+    StringView name; ///< everything up to (but excluding) the first slash('/')
+    StringView version; ///< everything after the name, including the slash('/')
+};
+
+std::ostream &operator <<(std::ostream &, const ProtocolView &);
+
+// HTTP is not explicit about case sensitivity of Upgrade protocol strings, but
+// there are bug reports showing different case variants used for WebSocket so
+// we preserve the received case but compare case-insensitively.
```

Comment:
RFC7230 defines the values as protocol names. The relevant RFC for each protocol defines what octets are to be used for those names.

RFC 6455 for WebSocket protocol explicitly defines it as lower case octets "websocket". The IANA registration section sadly used mixed-case "WebSocket" and there is an errata about this. So if we want to be tolerant we can accept the "WebSocket" variant in relaxed parser. Strict parser should only upgrade on the lower case. Any other case-insensitive match of "websocket" or "websockets" should be prevented from upgrading with a "SECURITY NOTICE" (its an attack underway).